sapi/fpm/fpm/fpm_stdio.c

```diff
@@ -151,49 +151,37 @@ static void fpm_stdio_child_said(struct fpm_event_s *ev, short which, void *arg)
 		log_stream = child->log_stream;
 	}
 
-	while (fifo_in || fifo_out) {
-		if (fifo_in) {
-			res = read(fd, buf + in_buf, max_buf_size - 1 - in_buf);
-			if (res <= 0) { /* no data */
-				fifo_in = 0;
-				if (res == 0 || (errno != EAGAIN && errno != EWOULDBLOCK)) {
-					/* pipe is closed or error */
-					read_fail = (res < 0) ? res : 1;
-				}
-			} else {
-				in_buf += res;
-				/* if buffer ends with \0, then the stream will be finished */
-				if (!buf[in_buf - 1]) {
-					finish_log_stream = 1;
-					in_buf--;
-				}
+	while (1) {
+		in_buf = read(fd, buf, max_buf_size - 1);
+		if (in_buf <= 0) { /* no data */
+			if (in_buf == 0 || (errno != EAGAIN && errno != EWOULDBLOCK)) {
+				/* pipe is closed or error */
+				read_fail = (in_buf < 0) ? in_buf : 1;
 			}
+			break;
 		}
 
-		if (fifo_out) {
-			if (in_buf == 0) {
-				fifo_out = 0;
-			} else {
-				char *nl;
-
-				nl = memchr(buf, '\n', in_buf);
-				if (nl) {
-					/* we should print each new line int the new message */
-					int out_len = nl - buf;
-					zlog_stream_str(log_stream, buf, out_len);
+		for (start = 0, pos = 0; pos < in_buf; pos++) {
+			switch (buf[pos]) {
+				case '\n':
+					zlog_stream_str(log_stream, buf + start, pos - start);
 					zlog_stream_finish(log_stream);
-					/* skip new line */
-					out_len++;
-					/* move data in the buffer */
-					memmove(buf, buf + out_len, in_buf - out_len);
-					in_buf -= out_len;
-				} else if (in_buf == max_buf_size - 1 || !fifo_in) {
-					/* we should print if no more space in the buffer or no more data to come */
-					zlog_stream_str(log_stream, buf, in_buf);
-					in_buf = 0;
-				}
+					start = pos + 1;
+					break;
+				case '\0':
+					if (pos + sizeof(FPM_STDIO_CMD_FLUSH) <= in_buf &&
```

Comment:
Shouldn't this be `<`?

Reply:
Nope, it's including the last '\0'. The rest of the logic should be like that for all other cases.

Reply:
Yes, it includes the last \0, but that \0 still need to be part of the buffer, which it isn't with a <= check.

Reply:
If buf contains only the separator, then in_buf is 6. It means that pos is 0 and sizeof(FMP_STDIO_CMD_FLUSH) is 6. So it's 0 + 6 <= 6. Am I missing something?

Reply:
No, you're right. My bad.

Reply:
Cool! Thanks for review anyway!